### PR TITLE
Add error handling for Azure Exceptions

### PR DESCRIPTION
**Pull request by sharonravindran**
Added error pipeline and message for Azure Connection Exceptions

### CODE REVIEW

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Contracts/Definition.cs [hunk 1]
```diff
@@ -14,5 +14,37 @@ public static readonly
...
+        /// <summary>
+        /// Error message 
+        /// </summary>
+        public string message { get; set; }
+    }
+
+    /// <summary>
+    /// Result object for Definition
+    /// </summary>
+    public class DefinitionResult
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Contracts/Definition.cs]: I'd move this out of the Contracts/Definition.cs file - probably create a new file at the LanguageServices folder level. The reasoning is that Contracts files map to data and events sent over the JSON-RPC layer. As this class is used internally and won't be sent on that channel, we shouldn't have it here.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Contracts/Definition.cs [hunk 2]
```diff
@@ -14,5 +14,37 @@ public static readonly
...
+    {
+        /// <summary>
+        /// True, if definition error occured
+        /// </summary>
+        public bool Error;
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Contracts/Definition.cs]: This should be HasError. Boolean properties should follow standard "IsXyz", "HasXyz", "DidXyz" patterns.

Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Contracts/Definition.cs]: You could also use "IsErrorResult" if you prefer.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 3]
```diff
@@ -54,6 +56,8 @@ private Database Database
                     catch(Exception ex)
                     {
                         Logger.Write(LogLevel.Error, "Exception at PeekDefinition Database.get() : " + ex.Message);
+                        this.error = true;
+                        this.errorMessage = SR.PeekDefinitionAzureError;
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: This is a dangerous assumption. Many different errors could have caused a general exception. We should only give this kind of message if we're sure.

I'd recommend 2 improvements:
- Always include the actual error message back to the user. The default error should be something like string.Format("An unexpected error occurred during Peek Definition execution: '{0}', ex.Message);
- Catch the most specific exception you can, e.g. ConnectionFailureException. Then _only_ if the target is Azure should you return this message. And you should still return the actual error message too unless you're 100% sure of the cause.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 4]
```diff
@@ -163,17 +167,23 @@ internal Location[] GetScript(IEnumerable<Declaration> declarationItems, string
                         // This workaround ensures that a schema name is present by attempting 
                         // to get the schema name from the declaration item 
                         // If all fails, the default schema name is assumed to be "dbo"
-                        if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && string.IsNullOrEmpty(schemaName))
+                        if (connectionInfo.ConnectionDetails.AuthenticationType.Equals("SqlLogin") && string.IsNullOrEmpty(schemaName))
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: SqlLogin should be a constant (in a Constants.cs file somewhere) given it's used around the codebase. Can you fix?

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs (hunk 3 above)
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: Modified

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs [hunk 5]
```diff
@@ -723,27 +731,57 @@ internal Location[] GetDefinition(TextDocumentPosition textDocumentPosition, Scr
...
+                                Locations = null
+                            };
                         });
 
                     // wait for the queue item
                     queueItem.ItemProcessed.WaitOne();
-                    return queueItem.GetResultAsT<Location[]>();
+                    return queueItem.GetResultAsT<DefinitionResult>();
+                }
+                catch (Exception ex)
```
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: This is to catch any exceptions in the Intellisense call. PeekDefinition handles its exceptions separately

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs [hunk 6]
```diff
@@ -48,6 +48,8 @@ public sealed class LanguageService
 
         internal const int PeekDefinitionTimeout = 10 * OneSecond;
 
+        internal static ServiceHost ServiceHost= ServiceHost.Instance;
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: You don't use this at all, as far as I can tell. Shouldn't it be removed?

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 7]
```diff
@@ -7,12 +7,13 @@
...
 using System.Collections.Specialized;
 using System.Data.SqlClient;
-using System.Runtime.InteropServices;
 using Microsoft.SqlServer.Management.Smo;
 using Microsoft.SqlServer.Management.Common;
 using Microsoft.SqlServer.Management.SqlParser.Intellisense;
 using Microsoft.SqlTools.ServiceLayer.Utility;
 using Microsoft.SqlTools.ServiceLayer.Connection;
+using Microsoft.SqlTools.ServiceLayer.Hosting.Protocol;
+
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: Please remove empty line

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 8]
```diff
@@ -62,20 +66,29 @@ private Database Database
...
-                            Server server = new Server(peekConnection);                        
-                            this.database = new Database(server, peekConnection.DatabaseName);                        
+                            Server server = new Server(peekConnection);
+                            this.database = new Database(server, peekConnection.DatabaseName);
+                        }
+                        catch (ConnectionFailureException cfe)
+                        {
+                            Logger.Write(LogLevel.Error, "Exception at PeekDefinition Database.get() : " + cfe.Message);
+                            this.error = true;
+                            this.errorMessage = connectionInfo.IsAzure ? SR.PeekDefinitionAzureError(cfe.Message) : SR.PeekDefinitionError(cfe.Message); ;
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: You never validate connectionInfo is not null. Please check here since it's possible (in the tests there are multiple places you don't define it, in which case this would throw null ref exception)

File: test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs [hunk 9]
```diff
@@ -231,23 +243,27 @@ public void GetTableDefinitionInvalidObjectTest()
...
         }
 
         /// <summary>
-        /// Test GetDefinition with an unsupported type(function)
+        /// Test GetDefinition with an unsupported type(function). Expect a error result.
         /// </summary>
         [Fact]
-        public void GetUnsupportedDefinitionForFullScript()
+        public void GetUnsupportedDefinitionErrorTest()
         {
```
Comment (kevcunnane) [inline test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs]: Minor: remove whitespace lines at top & bottom of method

Comment (kevcunnane) [inline test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs]: Minor: please add a `// given I am querying an unsupported object (Schema)` to match given/when/then pattern and make clear that it's the dbo schema you are getting the definition for

File: test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs [hunk 10]
```diff
@@ -257,27 +273,112 @@ public void GetUnsupportedDefinitionForFullScript()
...
+            // When I call the language service
+            var result = languageService.GetDefinition(textDocument, scriptFile, connInfo);
+
+            // Then I expect null locations and an error to be reported
+            Assert.NotNull(result);
+            Assert.True(result.IsErrorResult);
+
+        }
+
+
```
Comment (kevcunnane) [inline test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs]: Minor: remove unnecessary whitespace (2 empty lines instead of 1)

File: test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs [hunk 11]
```diff
@@ -257,27 +273,112 @@ public void GetUnsupportedDefinitionForFullScript()
...
+            string objectName = "from";
+
+            List<Declaration> declarations = new List<Declaration>();
+            DefinitionResult result = peekDefinition.GetScript(declarations, objectName, null);
+
+            Assert.NotNull(result);
+            Assert.True(result.IsErrorResult);
+            Assert.Equal(SR.PeekDefinitionNoResultsError, result.Message);
+        }
+
```
Comment (kevcunnane) [inline test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs]: Minor: remove unnecessary whitespace (2 empty lines instead of 1)

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs (hunk 6 above)
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: Yes, it should be removed. I added this while fixing the tests.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 12]
```diff
@@ -203,9 +225,9 @@ internal string GetSchemaFromDatabaseQualifiedName(string fullObjectName, string
             string[] tokens = fullObjectName.Split('.');
             for (int i = tokens.Length - 1; i > 0; i--)
             {
-                if(tokens[i].Equals(objectName))
+                if (tokens[i].Equals(objectName))
                 {
-                    return tokens[i-1];
+                    return tokens[i - 1];
```
Comment (llali) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: does the name always has schema? shouldn't you check for when tokens has only one item in it?

Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: If it has only one item then there isn't a schema name. This was a workaround for the SMO DefaultSchema bug

Comment (llali) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: ok so my problem is that you're getting tokens[i - 1] without checking if the tokens has more than item 

---
In reply to: [92512474](https://github.com/Microsoft/sqltoolsservice/pull/177#discussion_r92512474) [](ancestors = 92512474)
